runtime/v2/runc/task/service.go

```diff
@@ -517,48 +523,63 @@ func (s *service) send(evt interface{}) {
 	s.events <- evt
 }
 
-func (s *service) sendL(evt interface{}) {
-	s.eventSendMu.Lock()
-	s.events <- evt
-	s.eventSendMu.Unlock()
-}
-
 func (s *service) checkProcesses(e runcC.Exit) {
 	s.mu.Lock()
 	defer s.mu.Unlock()
 
+	missing := make(map[process.Process]*runc.Container)
 	for _, container := range s.containers {
-		if !container.HasPid(e.Pid) {
-			continue
-		}
-
 		for _, p := range container.All() {
+			if _, starting := s.processEventsLocker.Load(p.ID()); starting {
+				missing[p] = container
+				continue
+			}
 			if p.Pid() != e.Pid {
 				continue
 			}
 
-			if ip, ok := p.(*process.Init); ok {
-				// Ensure all children are killed
-				if runc.ShouldKillAllOnExit(s.context, container.Bundle) {
-					if err := ip.KillAll(s.context); err != nil {
-						log.L.WithError(err).WithField("id", ip.ID()).
-							Error("failed to kill init's children")
-					}
-				}
+			s.sendProcessEvent(e, p, container)
+			return
+		}
+	}
+
+	tryMissing := func(e runcC.Exit, p process.Process, c *runc.Container) {
+		for {
+			if _, starting := s.processEventsLocker.Load(p.ID()); !starting {
+				break
 			}
+			time.Sleep(50 * time.Millisecond)
+		}
 
-			p.SetExited(e.Status)
-			s.sendL(&eventstypes.TaskExit{
-				ContainerID: container.ID,
-				ID:          p.ID(),
-				Pid:         uint32(e.Pid),
-				ExitStatus:  uint32(e.Status),
-				ExitedAt:    protobuf.ToTimestamp(p.ExitedAt()),
-			})
+		if p.Pid() != e.Pid {
 			return
 		}
-		return
+		s.sendProcessEvent(e, p, c)
+	}
+	for p, c := range missing {
+		go tryMissing(e, p, c)
```

Comment:
The event channel is closed on shutdown and shutdown is protected by the same `s.mu` lock. Is completing this event outside of the lock an important part of reducing the contention issue? If not, its probably safest just to ensure these complete before returning.

Reply:
The most relevant change to resolving the contention issue is not blocking `checkProcesses` on processes that haven't started yet – so ensuring these complete before returning isn't a great solution and reintroduces a big part of the issue, but we can move around where we lock on `s.mu` to make sure we don't send events outside of it.

Reply:
The channel close/send behavior is pretty annoying in Go as it necessitates additional locking and status checking to avoid a possible panic. It might warrant an explicit `shutdown` state variable and mutex for send/close.